### DISCUSSION TITLE
Introduce rule for cyclomatic complexity

### DISCUSSION THREAD

**Discussion by arendjr** [rome/tools#4020] [Suggestions]
I would like Rome to have a lint rule that can enforce a maximum cyclomatic complexity.

At my work we heavily rely on this rule with ESLint in order to flag when it’s time to refactor/split a function because it’s becoming too complex.

Prior art:
* https://github.com/SonarSource/eslint-plugin-sonarjs/blob/HEAD/docs/rules/cognitive-complexity.md
* https://eslint.org/docs/latest/rules/complexity

**Comment by xunilrj** [comment]
Is this the paper describing "Cognitive Complexity"? https://www.sonarsource.com/docs/CognitiveComplexity.pdf

**Comment by arendjr** [comment]
Yes, I believe that’s the one the SonarJS implementation is based on. It suggests a number of improvements over simple cyclomatic complexity, as used by the stock ESLint rule.

**Comment by arendjr** [comment]
It's not yet done, but I've started implementation of this in my branch: https://github.com/arendjr/rome-tools/tree/cognitive-complexity

**Reply by arendjr** [reply]
Update: the PR is here: https://github.com/rome/tools/pull/4657

**Comment by arendjr** [comment]
Rule is merged :)
